Add gas configuration support to SignerLock

jcortejoso opened the pull request:
Introduce a new GasConfig struct to manage gas fees for L1 transactions, allowing for optional max fee and priority fee settings:
- Update SignerLock to include gas configuration, with methods for creating instances from environment variables.
- Modify the send_transaction_request method to apply gas settings when sending transactions.
- Add unit tests for GasConfig and SignerLock to ensure correct behavior and defaults.

NOTE: PR coded with cursor.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds an adaptive EIP-1559 gas pricing system with env-configurable overrides/caps and RBF retries, integrated into `SignerLock`, plus extensive tests and a `tracing` dependency.
> 
> - **utils/signer**:
>   - **Adaptive Gas Pricing**:
>     - Add `AdaptiveGasConfig` (with `from_env`) supporting overrides (`MAX_FEE_PER_GAS`, `MAX_PRIORITY_FEE_PER_GAS`), thresholds, history window, multiplier, percentile, caps (`MAX_GAS_PRICE_CAP`), skip threshold (`SKIP_IF_GAS_ABOVE`), and RBF settings.
>     - Add `AdaptiveGasOracle` to compute EIP-1559 fees via `eth_feeHistory`, percentile-based smoothing, timeout escalation, and apply to `TransactionRequest`; supports confirmation notifications and submission skipping.
>     - Add `GasPricingStrategy` (`Default` | `Adaptive`) with env-driven selection.
>   - **SignerLock**:
>     - Store `gas_strategy`; new constructors (`new_with_adaptive_gas`, `new_with_strategy`) and accessors (`gas_strategy`, `gas_oracle`, `gas_config`).
>     - Update `send_transaction_request` to apply oracle-estimated fees, fetch and pin nonce, and run RBF loop with configurable bump and cap enforcement.
>   - **Support types**: Add `GasPriceEstimate` and percentile helpers.
> - **Tests**: Add comprehensive unit tests for env parsing, percentile calc, EIP-1559 invariants, RBF bumping, and strategy wiring.
> - **Deps**: Add `tracing` to `utils/signer`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 523224d157092e0fd09e2ec83774963647fde040. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

jcortejoso committed:
Add gas configuration support to SignerLock

Introduce a new GasConfig struct to manage gas fees for L1 transactions, allowing for optional max fee and priority fee settings. Update SignerLock to include gas configuration, with methods for creating instances from environment variables. Modify the send_transaction_request method to apply gas settings when sending transactions. Add unit tests for GasConfig and SignerLock to ensure correct behavior and defaults.

palango commented:
My understanding is that gas values are added to the transaction in the `GasFiller`: https://docs.rs/alloy/latest/alloy/providers/fillers/struct.GasFiller.html

The alternative approach would be to create a custom one and inject that into the alloy provider.

jcortejoso commented:
> My understanding is that gas values are added to the transaction in the `GasFiller`: https://docs.rs/alloy/latest/alloy/providers/fillers/struct.GasFiller.html
> 
> The alternative approach would be to create a custom one and inject that into the alloy provider.

My understanding was that to customize the gas configuration for a transaction, you must use the [TransactionRequest](https://docs.rs/alloy/latest/alloy/rpc/types/eth/struct.TransactionRequest.html) struct. If you set there related fields (`gas_price`, `max_fee_per_gas`, `max_priority_fee_per_gas`, etc...) then the `GasFiller` don't fill that. If the gas config is missing then `GasFiller` is responsable of filling those values.

jcortejoso committed:
Refactor GasConfig to improve environment variable parsing

Enhance the GasConfig struct's from_env method to return a Result, allowing for error handling when environment variables are invalid. Introduce a new helper function, parse_env_var, to streamline the parsing process and provide detailed error messages. Update tests to cover various scenarios, including valid, partial, and invalid environment variable values.

jcortejoso committed:
Implement adaptive gas pricing strategy in SignerLock

Introduce AdaptiveGasConfig and AdaptiveGasOracle to manage gas pricing based on network conditions. Update SignerLock to support adaptive gas pricing, allowing for dynamic fee adjustments when gas prices exceed a specified threshold. Enhance the gas pricing strategy to include options for static and adaptive configurations, with corresponding methods for creating instances from environment variables. Update tests to validate the new functionality and ensure correct behavior of the adaptive gas pricing logic.

gastonponti commented on utils/signer/src/lib.rs:
```diff
@@ -187,38 +194,549 @@ impl Signer {
+        if let Some(timeout) = Self::parse_env_var_u64("GAS_FALLBACK_TIMEOUT_MINUTES")? {
+            config.fallback_timeout_minutes = timeout;
+        }
+        if let Some(percent) = Self::parse_env_var_u64("GAS_FALLBACK_INCREASE_PERCENT")? {
```
Shouldn't be also be a float? 2.5% should be valid too

gastonponti commented on utils/signer/src/lib.rs:
```diff
@@ -187,38 +194,549 @@ impl Signer {
+        Ok(config)
+    }
+
+    fn parse_env_var_u128(var_name: &str) -> Result<Option<u128>> {
```
This could be done with some generics, to avoid repetition. Also the parse_env that is used above for the max_fee_per_gas and the priority one

jcortejoso committed:
Refactor gas pricing configuration in Signer

Remove the legacy GasConfig struct and enhance AdaptiveGasConfig to include optional max fee and priority fee overrides. Update the from_env method to support parsing these overrides from environment variables, allowing for both static and adaptive gas pricing strategies. Modify SignerLock to utilize the new configuration structure, ensuring correct application of gas settings during transaction requests. Update tests to validate the new behavior and ensure compatibility with the adaptive pricing logic.

gastonponti commented on utils/signer/src/lib.rs:
```diff
@@ -187,38 +194,549 @@ impl Signer {
+    ///
+    /// This method fetches current gas prices and historical fee data to
+    /// calculate optimal gas prices based on the configured strategy.
+    pub async fn get_gas_prices<P>(&self, provider: &P) -> Result<GasPriceEstimate>
```
Maybe we could use only the current_gas_price and avoid the historic part. Ask for the gasPrice, wait, and ask the gasPrice again, until you have something under the threshold, or you've waited enough

jcortejoso replied:
I think we preferred something more heuristic. Having a static max baseFee cap can be difficult to balance between not delaying too much with high gas pressure and don't spend too much on isolated gas peaks

gastonponti replied:
Ok, I was thinking in keeping it as simple as possible. But I'm not against it

jcortejoso committed:
Enhance environment variable parsing in AdaptiveGasConfig

Refactor the parsing logic in AdaptiveGasConfig to use a generic parse_env_var function, improving error handling and providing clearer messages for invalid inputs. Update the from_env method to ensure all gas pricing configuration values are validated, including checks for finite positive numbers. Add tests to verify the rejection of invalid values such as NaN and zero for critical parameters.

jcortejoso committed:
Enhance AdaptiveGasConfig with new gas pricing features

Add optional fields to AdaptiveGasConfig for maximum gas price cap, skip submission if gas exceeds a threshold, and percentile-based fee calculations. Update the fallback increase percentage to use f64 for more precise values. Refactor the gas pricing logic in AdaptiveGasOracle to utilize these new configurations, including RBF (Replace-By-Fee) settings for transaction retries. Extend tests to cover the new configuration options and ensure correct behavior in various scenarios.

jcortejoso committed:
Refactor AdaptiveGasConfig and SignerLock for improved gas pricing handling

Update the from_env method in AdaptiveGasConfig to initialize max fee and priority fee overrides directly, enhancing clarity and reducing redundancy. Modify GasPricingStrategy to use Box for AdaptiveGasOracle, ensuring proper memory management. Adjust SignerLock to accommodate these changes, including updates to gas_oracle method for better reference handling. Add comments to clarify locking behavior during nonce fetching to prevent race conditions.

jcortejoso committed:
Clamp percentile values in calculate_percentile method to prevent out-of-bounds access. Add unit tests to verify clamping behavior for various edge cases, including out-of-range and special floating-point values.

jcortejoso committed:
Update AdaptiveGasConfig and GasPricingStrategy to enhance gas pricing configuration

- Changed default value for skip_if_gas_above to match max_gas_price_cap for consistency.
- Expanded environment variable checks in GasPricingStrategy to include new gas pricing parameters.
- Added unit tests to validate behavior when environment variables for gas pricing are missing or set.

jcortejoso committed:
Fix Cargo fmt

jcortejoso committed:
Adding tracing dependency

jcortejoso committed:
Merge branch 'develop' of github.com:celo-org/op-succinct into jcortejoso/gas

jcortejoso committed:
Merge branch 'develop' of github.com:celo-org/op-succinct into jcortejoso/gas

jcortejoso committed:
Enhance AdaptiveGasConfig and AdaptiveGasOracle with new EIP-1559 fee management features

- Added validation to ensure RBF_TIMEOUT_SECONDS is greater than 0.
- Introduced helper functions to clamp priority fees to max fees and ensure max fees are greater than or equal to priority fees.
- Updated gas price estimation logic to respect new fee management rules.
- Implemented RBF bumping logic that respects a maximum gas price cap.
- Added unit tests for new fee management functions and RBF bumping behavior.

jcortejoso committed:
Fix Cargo fmt
